### ISSUE TITLE
Is it possible for a live data update?

### ISSUE THREAD

**Issue by nextglory** [fooplugins/FooTable#422]
In terms of the data load, there are majorly 2 ways:
1) Cache a huge list of data to client and paginate it
2) Only fetch the page data when requested.

Personally, I prefer the 2) way, because it's not restricted by data count(no matter how big the rows are), and a great bandwidth saver.  No user likely will click and view all the pages one by one.

If FooTable supports this feature in future, it will be great, user can choose between the client/live modes, for the 2), just give a rows total number, it will render the pagination, but only load one page at a time.

@steveush , what do you think?

Andy

**Comment by steveush** [comment]
Hi Andy,

In response to the title question: no at the moment it's not possible unless you build it custom.

I also prefer approach (2) and this was actually how the first beta versions of V3 were implemented. Please take a look at my answer to a [similar question](https://github.com/fooplugins/FooTable/issues/402) for the reasoning as to why it was changed to approach (1). To sum it up it was too difficult for people to implement as it required them to write endpoints and database queries in the server languages of their choice so we went with the easiest approach where we could handle everything client side.

As I said in that previous question it is something I will be looking at to add back into the plugin in some form or another but my time for this project is limited.

Thanks

**Comment by nextglory** [comment]
Thanks, @steveush ,

I got your point, as a opensource project, an easiest and shortest learning curve is important.
If choose to build it custom,  core/paging component can be possibly modified to have that feature? or **What's the starting point from your suggestion**?

**Comment by steveush** [comment]
Hey Andy,

One thing to remember is that you couldn't just handle paging server side. You would have to handle filtering and sorting as well as these two components will have an effect on what should be returned for any single page. It's the reason those are the three core components, they work very tightly together when it comes to what is displayed in the UI.

The best place to start would be looking at the how the `Table` object drives the execution of the child component methods. Basically how the `Table#draw` method loops through all registered components and executes the `predraw`, `draw` and `postdraw` methods on each one. Once you see how that all works the idea behind adding in ajax functionality becomes the following:

(1) Add a new option called `ajax`. This option when provided should be a function that returns a Promise (jQuery.Deferred) object so the plugin can wait for it's execution to complete. The function provided here is what would actually perform the ajax request to the server and return the results and could be as simple as the following seeing as jQuery.ajax returns a Promise.

``` javascript
$('#table-id').footable({
    "ajax": function(request){
        return $.ajax('endpoint', {data: request});
    }
});
```

(2) Extend the base `Component` object to have a `preajax` and `postajax` methods that can be hooked into by any component. The `preajax` method would allow components to build up an object containing values to send to the server, think of it as a request object, for example:

``` javascript
{
    "paging": {
        "page": 1, // The page to retrieve
        "size": 10 // The size of the page
    },
    "filtering": {
        "filters": [] // An array of filters to apply to the request
    },
    "sorting": {
        "column": "some_column", // The column to sort on
        "direction": "ASC" // The direction to sort it
    }
}
```

Using the request object you can build up the query to execute against your data store.

The `postajax` method would allow you to handle the response from the server for each component. An example response could be something as simple as the below:

``` javascript
{
    "paging": {
        "total": 10000 // The total number of pages
    },
    "rows": [] // An array of row objects
}
```

The response would contain the total number of pages available on the server (this would allow us to generate the pagination links) as well as the row objects returned by executing the request object against your database.

(3) Create a new method in the `Table` object, similar to the `Table#draw` method, to drive the execution of the ajax functionality, for purposes of this explaination I'll call it `Table#ajax`. It should first check if the `ajax` option is provided, if it is then execute the `preajax` method to build the request object. It would then execute the `ajax` option function passing in the built up request object and wait for it to be resolved. Once resolved it takes the response and executes the `postajax` method allowing each component to do as it needs with the results. Once that is done it then executes the `Table#draw` method allowing the table to redraw itself with the updated values.

(4) Modify each of the components as required to handle the `preajax` and `postajax` methods. You would have to also modify the `Rows` component to correctly update it's internal workings with the new row objects returned from the server.

(5) Modify the components to call the `Table#ajax` method when required. For example with paging you would want to call it whenever a pagination link to a different page is clicked.

The above is a very high level overview of what would be required. There would need to be work done in each component to make sure it could handle the ajax response and update itself.

Thanks
